aten/src/ATen/native/sparse/ValidateCompressedIndicesCommon.h

```diff
@@ -0,0 +1,285 @@
+#pragma once
+
+#include <ATen/Tensor.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/Dispatch.h>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/arange.h>
+#include <ATen/ops/empty.h>
+#include <ATen/ops/_sparse_coo_tensor_with_dims_and_tensors.h>
+#endif
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+#define GPUCC
+#define FUNCAPI __host__ __device__
+#define INLINE __forceinline__
+#define NAME "compressed_index_invariance_checks_cuda"
+#else
+#define FUNCAPI
+#define INLINE inline
+#define NAME "compressed_index_invariance_checks_cpu"
+#endif
+
+#if defined(_WIN32) || defined(_WIN64)
+#define RESTRICT __restrict
+#else
+#define RESTRICT __restrict__
+#endif
+
+#define INVARIANT_CHECK_FUNC_API static INLINE FUNCAPI void
+
+namespace at {
+namespace native {
+
+namespace {
+
+// NOTE: all the checks but the very last one are designed
+// to work with vectors.
+// To enable vectorization one would need to write a conversion
+// Vec -> bool and make kernel launchers call into vectorized
+// execution paths.
+
+// All the invariants are described in https://pearu.github.io/bsr_tensor_invariants.html
```

Comment:
While they're written out on the website, could we also copy-paste the content into here so that users don't need to cross-reference while reading this code?

Reply:
There are listed in the comments below.